circularqueue.gen.go.tmpl

```diff
@@ -0,0 +1,111 @@
+//lint:file-ignore U1000 this is generated code
+package kapacitor
+
+{{with $types := .}}{{range $k := $types}}
+
+// {{ $k }}CircularQueue defines a circular queue, always use the contructor to create one.
+type {{ $k }}CircularQueue struct {
+	data     []{{ $k }}
+	head     int
+	tail     int
+	l        int
+}
+
+
+// {{ if eq (substr 0 1 $k ) (substr 0 1 $k | upper) -}} New {{- else -}} new {{- end -}}
+{{- substr 0 1 $k | upper -}}{{- substr 1 (len $k) $k -}} constructs a Circular Queue
+// with a given buffer buf.  It is ok for buf to be nil.
+func {{ if eq (substr 0 1 $k ) (substr 0 1 $k | upper) -}} New {{- else -}} new {{- end -}}
+{{- substr 0 1 $k | upper -}}{{- substr 1 (len $k) $k -}} CircularQueue(buf ...{{ $k }}) *{{ $k }}CircularQueue {
+	// if we have a useless buffer, make one that is at least useful
+	if cap(buf) < 4{
+		buf = append(make([]{{ $k }}, 0, 4), buf...)
+	}
+	return &{{ $k }}CircularQueue{
+		data: buf[:cap(buf)],
+		tail: len(buf), // tail is here we insert
+		l:    len(buf),
+	}
+}
+
+// Enqueue adds an item to the queue.
+func (q * {{- $k -}} CircularQueue) Enqueue(v {{ $k }}) {
+	// if full we must grow and insert together. This is an expensive op
+	if cap(q.data) > q.l {// no need to grow
+		if q.tail == len(q.data){
+			q.tail = 0
+		}
+		q.data[q.tail] = v
+	}else{ // we need to grow
+		buf := make([]{{ $k }}, cap(q.data)*2)
+		if q.head < q.tail{
+			copy(buf, q.data[q.head:q.tail])
+		} else {
+			copy(buf[copy(buf, q.data[q.head:]):], q.data[:q.tail])
```

Comment:
This might be easier to understand written on more than one line:

```
partialLength := copy(buf, q.data[q.head:])
copy(buf[partialLength):], q.data[:q.tail])
```